proposals/0118-partitioned-epoch-reward-distribution.md

```diff
@@ -0,0 +1,255 @@
+overflowing the number of blocks in an epoch, the number of blocks is capped at
+10% of the total slots in an epoch.
+
+The [SipHash 1-3](https://www.aumasson.jp/siphash/siphash.pdf) pseudo-random
+function is used to hash stake account addresses efficiently and uniformly
+across the blocks in the reward distribution phase. The hashing function for an
+epoch is created by seeding a new `SipHasher` with the parent block's blockhash.
+This hashing function can then be used to hash each active stake account's
+address into a u64 hash value. The reward distribution block index `I` can then
+be computed by applying the following formula to the hash:
+
+```
+I = (M * stake_address_hash) / 2^64
+```
+
+### Reward Distribution
+
+The reward distribution phase happens after the reward computation phase, ie.
+during the second block in the epoch as per the current design. It lasts for `M`
+blocks (see previous section).
+
+#### `EpochRewards` Sysvar Account
+
+The `EpochRewards` sysvar account records whether the rewards distribution phase
+is in progress, as well as the details needed to resume distribution when
+starting from a snapshot during the reward distribution phase. These details
+include: the parameters needed to recalculate the reward partitions, the total
+rewards to be distributed, and the amount of rewards distributed so far.
+
+The layout of `EpochRewards` sysvar is shown in the following pseudo code.
+
+```
+#[repr(C)] // C representation, ie. 8-byte alignment
+struct EpochRewards{
+   // whether the rewards period (calculation and distribution) is active
+   active: bool, // byte. false: 0x00, true: 0x01
+
+   distribution_starting_block_height: u64, // little-endian unsigned 64-bit integer
+   
+   num_partitions: u64, // little-endian unsigned 64-bit integer
+   
+   parent_blockhash: Hash, // byte-array of length 32
+
+   // total rewards for the current epoch, in lamports
+   total_rewards: u64, // little-endian unsigned 64-bit integer
+
+   // distributed rewards for the current epoch, in lamports
+   distributed_rewards: u64, // little-endian unsigned 64-bit integer
+}
+```
+
+The `EpochRewards` sysvar is repopulated at the start of the first block of the
+epoch (before any transactions are processed), as both the total epoch rewards
+and vote account rewards become available at this time. The
+`distributed_rewards` field is updated per reward distribution for each block in
+the reward distribution phase. The `active` field is set to false after the last
+partition is distributed (and `total_rewards == distributed_rewards`), ie.
+before processing transactions in block at height `X + M`.
+
+#### Booting from Snapshot
+
+When booting from a snapshot, a node must check the EpochRewards sysvar account
+to determine whether the distribution phase is active. If so, the node must
+rerun the rewards calculation using the `EpochRewards::num_partitions` and
+`EpochRewards::parent_blockhash` sysvar fields as well as the `EpochStakes` in
```

Comment:
`EpochStakes` doesn't store the full stake account state. It only stores the stake delegation info, which is not enough to recompute the rewards. 
To recompute stake rewards, we have to store the full stake account state, which is as expensive as storing the rewards.
Therefore, I think we should keep the current implementation of storing the computed rewards in the snapshot.

Reply:
> we are guaranteeing that no state will have changed since the first block in the epoch

No, this is not true. When the stake account received rewards, its state changes.

Reply:
Well, correct. Sorry, I left that part implied. Explain how that is an issue, please.

Reply:
For example, the `observed_credit` field will change after receiving rewards. 
Reward computation will depend on that field. If we want to recompute rewards later, we need to know the `observed_credit` for all stake accounts at the epoch boundary.

Reply:
That field is inside stake `state`. but not saved in `EpochStake`. When we load the stake account later, we can't recover the old state.

Reply:
> that will add extra complexity and make startup from snapshot slower.

The motivation of this amendment is that a little extra time booting from snapshot is preferable to adding 100s of MBs to the snapshot.
I would expect the performance impact of that subtraction to be nominal, but if it is more than that, we definitely don't need to add each entire StakeState to the snapshot to preserve just `credits_observed`.

Reply:
There is also code complexity and risk aspect to consider. There are a lot of if-else and corner case handling in reward calculation code. If not careful, we could easily introduce a bug. And a bug there will be very bad. 

In my opinion, the extra space in snapshot doesn't sound too bad. 
1. We only pay the cost when the snapshot is taken during reward period. For snaphot take out of reward period, we don't need to pay the cost.
2. I think the reward data should have many similar data in the fields, which will make it very compressible. After compress, it woun't take 100s of MBs.
3. It makes startup from snapshot faster.

Reply:
I have simulated a bank snapshot with 1M stake rewards, randomly delegated to 2K vote accounts. 

Total space for uncompressed bank snapshot took 296M, compressed space 65M. 

```
-rw-r--r-- 1 sol users 296M Feb 24 00:38 foo.snap
-rw-r--r-- 1 sol users  65M Feb 24 00:38 foo.snap.lz4
```

Reply:
it's only partially about the space. adding snapshot fields must be adopted by all implementation. the fd team was lamenting about this at the offsite and we committed to avoid adding them whenever possible, instead preferring to store data in accounts. storing _this_ data in an account is a nonstarter though.

i don't think we even need to unwind the credits, but we might need to store a couple more fields in the sysvar. we already have the total and distributed lamports. if we also add the total and distributed credits, we can can trivally recover the distributions for the remaining partitions.

Reply:
I can argue that there is no free lunch. If you don't want to deal with the complexity of handling the new field, you will have to introduce the complexity of a new code path to recompute rewards. And you will have to maintain both reward code paths in future.  Besides, the field is already in v1.17. If fd, use v.1.17 and above as the reference impl, it would already have this field.

Reply:
Ofc we can ask, but I think it is highly unlikely that firedancer supports partitioned epoch rewards in any fashion yet. The fact that something is already implemented does not make it the right design, especially since the existing approach was not described (approved) in the original simd, nor was that simd signed off on by the firedancer team.

Reply:
Heh. What's the policy for simd approval, consensus or majority? Anyway, if I need to vote, my vote is against recomputing. But if the majority thinks that recomptuing is the way to go, I am fine to accept it too. [(*^ ‿ ⁢*](https://www.emoticonstext.com/wink.html)